src/Microsoft.AspNetCore.Sockets.Client/WebSocketsTransport.cs

```diff
@@ -0,0 +1,122 @@
+using System.Collections.Generic;
+
+namespace Microsoft.AspNetCore.Sockets.Client
+{
+    public class WebSocketsTransport : ITransport
+    {
+        private ClientWebSocket _webSocket = new ClientWebSocket();
+        private IChannelConnection<Message> _application;
+        private CancellationToken _cancellationToken = new CancellationToken();
+
+        public Task Running { get; private set; }
+                                                
+        public async Task StartAsync(Uri url, IChannelConnection<Message> application)  
+        {
+            if (url == null)
+            {
+                throw new ArgumentNullException(nameof(url));
+            }
+            if (application == null)
+            {
+                throw new ArgumentNullException(nameof(application));
+            }
+            _application = application;
+            var webSocketUrl = url.ToString();
+            await Connect(webSocketUrl);
+            var sendTask = SendMessages(url, _cancellationToken);
+            var receiveTask = ReceiveMessages(url, _cancellationToken);
+
+            Running = Task.WhenAll(sendTask, receiveTask).ContinueWith(t => {
+                _application.Output.TryComplete(t.IsFaulted ? t.Exception.InnerException : null);
+                return t;
+            }).Unwrap();
+        }
+
+        public async Task ReceiveMessages(Uri pollUrl, CancellationToken cancellationToken)
+        {
+            var totalBytes = 0;
+            var incomingMessage = new List<byte[]>();
+            while (!cancellationToken.IsCancellationRequested)
+            {
+                bool completedMessage;
+                WebSocketReceiveResult receiveResult;
+                ArraySegment<byte> buffer;
+                do
+                {
+                    buffer = new ArraySegment<byte>(new byte[1024]);
+                    receiveResult = await _webSocket.ReceiveAsync(buffer, cancellationToken);
+                    completedMessage = receiveResult.EndOfMessage;
+                    incomingMessage.Add(buffer.Array);
+                    totalBytes += buffer.Count;
+                } while (!completedMessage);
+
+                Message message;
+                if (incomingMessage.Count > 1)
+                {
+                    var totalBuffer = new byte[totalBytes];
+                    var offset = 0;
+                    for (int i = 0 ; i < incomingMessage.Count; i++)
+                    {
+                        System.Buffer.BlockCopy(incomingMessage[i], 0, totalBuffer, offset, incomingMessage[i].Length);
+                        offset += incomingMessage[i].Length;
+                    }
+
+                    message = new Message(ReadableBuffer.Create(totalBuffer).Preserve(), Format.Text, receiveResult.EndOfMessage);
+                }
+                else
+                {
+                    message = new Message(ReadableBuffer.Create(buffer.Array).Preserve(), Format.Text, receiveResult.EndOfMessage);
```

Comment:
I'd rather keep `buffer` scoped to the loop above. You can just use `incomingMessage[0]` if you really want to optimize the "one-buffer message" scenario.